src/Orleans/Threading/CancellationTokenHolderGrain.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Placement;
+
+namespace Orleans.Threading
+{
+    [PreferLocalPlacement]
+    internal class CancellationTokenHolderGrain : Grain, ICancellationTokenHolderGrain
+    {
+        private CancellationTokenSource cancellationTokenSource;
+        private bool collectionScheduled;
+
+        public override Task OnActivateAsync()
+        {
+            DelayDeactivation(TimeSpan.FromDays(5));
+            cancellationTokenSource = new CancellationTokenSource();
+            return TaskDone.Done;
+        }
+
+        public Task<CancellationToken> GetCancellationToken()
+        {
+            return Task.FromResult(cancellationTokenSource.Token);
+        }
+
+        public Task Cancel()
+        {
+            cancellationTokenSource.Cancel();
+            DelayDeactivation(TimeSpan.FromSeconds(30));
```

Comment:
I can imagine some edge case with extremely bad network which would requere multiple retries of message sending, that could span over 30 seconds,  with grain call timeout adjusted to more than 30 seconds. I could use the grain call timeout from the config as delay, but thought that probability of abovementioned case vanishingly low.